New package: ovmf

Francesco149 opened the pull request:
this is a uefi used in qemu virtual machines. it's automatically picked up by libvirtd and is quite popular for gpu passthrough setups

Duncaen commented:
I don't think its a good idea to source (`. edksetup.sh BaseTools`) stuff into the xbps-src environment.

Duncaen commented on srcpkgs/ovmf/template:
```diff
@@ -0,0 +1,47 @@
+		-t ${_toolchain_opt} \
+		-a X64 \
+		-p OvmfPkg/OvmfPkgX64.dsc \
+		-n $(nproc) \
```
`$XBPS_MAKEJOBS` instead of `nproc`.

Duncaen commented on srcpkgs/ovmf/template:
```diff
@@ -0,0 +1,47 @@
+		-D TLS_ENABLE
+}
+
+post_extract() {
```
ordering functions by execution order makes it easier to follow.

Francesco149 commented:
hm, what do you suggest for sourcing the environment, just do a subshell that sources it?

Francesco149 commented:
I'm gonna move it to a script in files

Duncaen commented:
Look at the alpine build, they just do `bash -c ". edksetup.sh"`.

Francesco149 commented:
ah nice, done

teldra commented:
maybe we can consolidate: https://github.com/void-linux/void-packages/pull/11392
This template can build ovmf X64 and IA32 for every arch (apart from i686, that only builds IA32)

Francesco149 commented:
oh, I totally missed that. one issue I see with that is that libvirtd only looks for `/usr/share/OVMF/OVMF_*.fd` , so placing multiple arch's in there would not work out of the box and you'd have to edit `/etc/libvirt/qemu.conf` . not a huge issue for me, but something to think about

Francesco149 commented:
it seems that the correct way to do it is to install metadata files that point to the firmware in /usr/share/qemu/firmware

example of a firmware metadata file https://src.fedoraproject.org/rpms/edk2/blob/master/f/50-edk2-ovmf-x64-sb.json

Francesco149 commented:
I've modified it to build x64 and ia32 as well as installing the proper json files. however this will not work properly without workarounds on VM creation until this bug is fixed: https://bugzilla.redhat.com/show_bug.cgi?id=1776949

teldra commented on srcpkgs/ovmf-ia32/template:
```diff
@@ -0,0 +1,102 @@
+maintainer="Franc[e]sco <lolisamurai@tfwno.gf>"
+license="BSD-2-Clause-Patent, MIT, Apache-2.0"
+homepage="https://github.com/tianocore/edk2/tree/master/OvmfPkg"
+_openssl_ver=1.1.1b
```
upstream switched to 1.1.1d because of CVE in 1.1.1b

teldra commented on srcpkgs/ovmf-ia32/template:
```diff
@@ -0,0 +1,102 @@
+	make -C BaseTools
+	# x86_64, no secure boot, no smm
+	command build -t ${_toolchain_opt} \
+		-a X64 -p OvmfPkg/OvmfPkgX64.dsc ${_ovmf_x86_flags}
```
64bit can not be buil on i686

teldra commented on srcpkgs/ovmf-ia32/template:
```diff
@@ -0,0 +1,102 @@
+		-a X64 -p OvmfPkg/OvmfPkgX64.dsc ${_ovmf_x86_flags}
+	# x86_64, secure boot, smm
+	command build -t ${_toolchain_opt} \
+		-a IA32 -a X64 -p OvmfPkg/OvmfPkgIa32X64.dsc ${_ovmf_x86_sb_flags}
```
why -a IA32  and -a X64?

Francesco149 replied:
it's what fedora's spec uses to build the secboot image. not entirely sure if that's required

teldra commented on srcpkgs/ovmf-ia32/template:
```diff
@@ -0,0 +1,102 @@
+archs=noarch
+wrksrc="edk2-edk2-stable${version}"
+hostmakedepends="make python acpica-utils nasm"
+makedepends="libuuid-devel"
```
this should be in `hostmakedepends`

Francesco149 commented:
looks good to me. every arch seems to be building it correctly

Duncaen commented on srcpkgs/ovmf-ia32/template:
```diff
@@ -0,0 +1,118 @@
+ -D TLS_ENABLE
+ -D FD_SIZE_2MB
+"
+_ovmf_x86_sb_flags="
```
why do they need to be toplevel variables.

Duncaen commented on srcpkgs/ovmf-ia32/template:
```diff
@@ -0,0 +1,118 @@
+		export HOST_ARCH=IA32
+		_ovmf_build_x64=false
+		;;
+esac
```
Why is `HOST_ARCH` exported on the toplevel, this leaks.

Usually we don't do `_ovmf_build_x64=false`, we either have it set or unset and then test if the variable is set `[ "$foo" ]` instead of making it a command that has to be executed, even if its builtin.

Duncaen commented on srcpkgs/ovmf-ia32/template:
```diff
@@ -0,0 +1,118 @@
+esac
+
+post_extract() {
+	mv "${XBPS_BUILDDIR}/openssl-${_openssl_ver}/"* CryptoPkg/Library/OpensslLib/openssl/
```
I'd rather have `create_wrksrc=yes` and have this selfcontained instead og leaking other things than `$wrksrc` into the builddir, this is not being correctly cleaned up because you only move files inside the `openssl-$x` directory.

Duncaen commented on srcpkgs/ovmf-ia32/template:
```diff
@@ -0,0 +1,118 @@
+# Template file for 'ovmf-ia32'
+pkgname=ovmf-ia32
```
I don't really like the invented ppackage name at the toplevel, can we just do it like alpine and package edk2 and make the ovmf stuff subpackages?

Francesco149 commented:
alpine seems to install the edk2 binaries in /usr/bin but some of them have very generic names like "build.py", so I decided to install in /usr/lib/edk2/bin . lemme know if you have better ideas.

anjandev commented:
It seems that the travis build failed because:

> The job exceeded the maximum log length, and has been terminated.

@Francesco149: Please follow the instructions here for huge packages: https://wiki.voidlinux.org/Xbps-src#Huge_packages

I just compiled on my local machine and I am running a windows vm using this package. Is there any reason this PR hasn't been merged? @Duncaen

Francesco149 commented:
@anjandev waiting for void's qemu version to catch up to the versions that fixes improper detection of the uefi binaries when creating a new vm

https://bugzilla.redhat.com/show_bug.cgi?id=1776949

Francesco149 commented:
hey, this should work as expected since we are now on libvirt-6.0.0. the uefi binaries detection was fixed in that release

anjandev commented:
@Duncaen can you please take a look at this PR again? I want to make a PCI passthrough guide for voidlinux and this PR being merged is a requirement.

q66 commented:
this kinda clusterfuck has no place being noarch, should be `archs="x86_64* i686*"`

and it's enough of a mess that I'd argue it should be `restricted=yes`

pullmoll commented on srcpkgs/edk2/template:
```diff
@@ -0,0 +1,165 @@
+		fulldir=BaseTools/Source/Python/${d}
+		vcopy ${fulldir} /usr/share/edk2/Python/
+		[ -f "${fulldir}/${d}.py" ] || continue
+		cat > PythonBins/${d}.py << EOF
```
Hint: with `<<- EOF` in the heredoc you can indent the lines to the same level as `cat > …` without them being indented in the output, i.e. leading tabs are removed.

Francesco149 commented:
hey sorry i haven't had time to look at this. so um, yes the toolchain should not be noarch. but the firmware blobs should, as they are used in virtual machines (which don't care about the host arch). I have updated the template to reflect that. It was left over from when I was just packaging the firmware blobs and not the toolchain. I've also updated the package to the latest version

let me know if i should make it `restricted=yes`

q66 commented:
these shouldn't be noarch either, generally nothing that's built conditionally should be noarch - and this is totally conditional, in both cases, since the template itself is conditional

we're planning to eventually drop support for noarch anyway

Francesco149 commented:
dropped the noarch's

q66 commented:
alright, since i don't do x86 i'll leave the rest to other reviewers

anjandev commented:
This pull request is blocking the pci passthrough guide. What else needs to be done for this to be merged?

Dragovorn commented:
Hello, I am encountering an issue when trying to build this package using `./xbps-src pkg ovmf-x64`, bear with me as I'm relatively new to using Void (and at most intermediate with Linux itself). I am encountering this error on build:  
```
=> ERROR: edk2-202005_1: vinstall: cannot find 'Build/OvmfX64NoSb/RELEASE_GCC5/FV/OVMF_CODE.fd'...
=> ERROR: edk2-202005_1: do_install: 'return 1' exited with 1
=> ERROR:   in _noglob_helper() at common/environment/setup/install.sh:12
=> ERROR:   in do_install() at srcpkgs/edk2/template:133
```
Upon investigation into `masterdir/builddir/edk2-202005/edk2-edk2-stable202005/Build/OvmfX64NoSb/RELEASE_GCC5/FV` I find that it is empty with only a `Ffs` directory present.

anjandev commented:
I was able to reproduce @Dragovorn 's issues

rinpatch commented:
The issue is due to a python 3.8/3.9 incompatibility, it is fixed in edk2 202011. I was able to successfully build the package with the following changes:
```diff
diff --git a/srcpkgs/edk2/template b/srcpkgs/edk2/template
index 0487bb7d34..d8571bd38b 100644
--- a/srcpkgs/edk2/template
+++ b/srcpkgs/edk2/template
@@ -1,6 +1,6 @@
 # Template file for 'edk2'
 pkgname=edk2
-version=202005
+version=202011
 revision=1
 archs="x86_64* i686*"
 create_wrksrc=yes
@@ -9,13 +9,13 @@ short_desc="EFI Development Kit II"
 maintainer="Franc[e]sco <lolisamurai@tfwno.gf>"
 license="BSD-2-Clause-Patent, MIT, Apache-2.0"
 homepage="https://github.com/tianocore/edk2/tree/master/OvmfPkg"
-_openssl_ver=1.1.1d
+_openssl_ver=1.1.1g
 _brotli_commit=666c3280cc11dc433c303d79a83d4ffbdd12cc8d
 distfiles="https://github.com/tianocore/edk2/archive/${pkgname}-stable${version}.tar.gz
  https://www.openssl.org/source/openssl-${_openssl_ver}.tar.gz
  https://github.com/google/brotli/archive/${_brotli_commit}.tar.gz"
-checksum="373c3eff3497316a48fcf4be8dcee227431cbce86dcd80a004950e992f0297e2
-1e3a91bc1f9dfce01af26026f856e064eab4c8ee0a8f457b5ae30b40b8b711f2
+checksum="53e877c714426c2c1957d1e875d99c7eb6602a858d7ec8ba4eac42c37e3ce064
+ddb04774f1e32f0c49751e21b67216ac87852ceb056b75209af2443400636d46
 cd82ce69defa40e9e36443e59d90aa5f8c6013024fae6a2de8c31afc2c707065"
 _toolchain_opt=GCC5
 
@@ -98,7 +98,7 @@ do_install() {
                vmkdir /usr/share/${pkgname}/${d}
        done
        vmkdir /usr/lib/${pkgname}/bin
-       bins="Brotli DevicePath EfiRom GenCrc32 GenFfs GenFv GenFw GenSec
+       bins="BrotliCompress DevicePath EfiRom GenCrc32 GenFfs GenFv GenFw GenSec
         LzmaCompress Split TianoCompress VfrCompile VolInfo"
        for bin in $bins; do
                vinstall BaseTools/Source/C/bin/${bin} 755 /usr/lib/${pkgname}/bin
```

Dragovorn commented:
Those changes made it build on my machine, thanks!

ericonr commented:
Closing this PR due to inactivity. Feel free to reopen.

@rinpatch, feel free to open your own, if you wish.

rinpatch commented:
Stopped using void since then, so it will have to be someone else.
